src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -167,18 +171,28 @@ public void Initialize()
                 }
 
                 _deadline = Clock.UtcNow.Add(timeout);
-                _cts = new CancellationTokenSource(timeout);
+
+                _timeoutCts = new CancellationTokenSource(timeout);
+                _timeoutCts.Token.Register(DeadlineExceeded, timeout);
+
+                _linkedCts = CancellationTokenSource.CreateLinkedTokenSource(_timeoutCts.Token, HttpContext.RequestAborted);
             }
             else
             {
                 _deadline = DateTime.MaxValue;
-                _cts = new CancellationTokenSource();
             }
         }
 
+        private void DeadlineExceeded(object timeout)
+        {
+            DeadlineExceeded(_logger, (TimeSpan)timeout);
+            HttpContext.Abort();
```

Comment:
I noticed some response headers that gRPC sets when the server his the deadline. Still need to set those. Still need to find out what happens when gRPC has started streaming and it hits the deadline.